### PR TITLE
fix: CSRF origin matching should be case-insensitive

### PR DESCRIPTION
**Pull request by jaffarkeikei**
## Summary

This PR fixes two bugs in the `isCsrfOriginAllowed` function used for Server Actions CSRF protection:

1. **Case sensitivity bug**: Domain matching was case-sensitive, but DNS names are case-insensitive per [RFC 1035](https://datatracker.ietf.org/doc/html/rfc1035#section-2.3.3). This caused legitimate requests to fail CSRF checks when the Origin header contained uppercase characters.

2. ~**Trailing dot bug**: FQDNs with trailing dots (e.g., `example.com.`) weren't matched. The trailing dot is valid DNS notation representing the root zone.~ Edit: Reverted, see comments on PR

## The Problem

Before this fix:
```javascript
isCsrfOriginAllowed('sub.VERCEL.com', ['*.vercel.com']) // false ❌
isCsrfOriginAllowed('sub.vercel.com.', ['*.vercel.com']) // false ❌
```

Both of these should return `true` because:
- DNS is case-insensitive (RFC 1035 §2.3.3)
- Trailing dots are valid FQDN notation

## The Fix

Normalize both domain and pattern before comparison:
- Convert to lowercase
- Strip trailing dots

## Testing

Added new test cases for:
- Case-insensitive matching (various case combinations)
- Trailing dot handling (in both domain and pattern)

All existing tests continue to pass.

## Related

- This affects users who configure `serverActions.allowedOrigins` in `next.config.js`
- The bug could cause legitimate Server Action requests to be rejected with CSRF errors if the browser sends an Origin header with different casing than the configured pattern

### CODE REVIEW

File: packages/next/src/server/app-render/csrf-protection.ts [hunk 1]
```diff
@@ -4,8 +4,12 @@
 // https://nextjs.org/docs/app/api-reference/components/image#remotepatterns
 // TODO - retrofit micromatch to work in edge and use that instead
 function matchWildcardDomain(domain: string, pattern: string) {
-  const domainParts = domain.split('.')
-  const patternParts = pattern.split('.')
+  // Normalize: DNS is case-insensitive (RFC 1035) and trailing dots are optional
+  const normalizedDomain = domain.toLowerCase().replace(/\.$/, '')
+  const normalizedPattern = pattern.toLowerCase().replace(/\.$/, '')
```
Comment (bgw) [inline packages/next/src/server/app-render/csrf-protection.ts]: How does this handle punycode domain names? Is it possible that this gets a unicode-encoded domain and tries to lowercase that, because that would be a security vulnerability as RFC 1035 only covers 7-bit ASCII. It's hard for me to know for certain if this function gets the punycode representation or the unicode representation.

You can address this in one of two ways:
- Write your own `toLowerCase` implementation that's careful to only modify A-Z.
- Add an e2e test to tests/e2e that proves this does not try to lower-case unicode characters. I'm not sure how you'd mock the DNS lookup though.

**Review by bgw — CHANGES_REQUESTED** [review]
At a high level: Yes, this looks like a good change.

I'm a bit concerned about unicode support though, so I'm requesting changes. See my inline comment.

**Comment by jaffarkeikei** [comment]
The test failures appear to be unrelated to the CSRF changes in this PR. All failing tests are showing `baseline-browser-mapping` deprecation warnings in stderr, which causes tests expecting empty stderr to fail.

The warning appears in multiple test suites:
- Production Custom Build Directory
- config-output-export
- webpack-require-hook  
- server-side dev errors

Would you like me to update `baseline-browser-mapping` to the latest version as part of this PR, or should this be handled separately?

File: packages/next/src/server/app-render/csrf-protection.ts (hunk 1 above)
Comment (jaffarkeikei) [reply packages/next/src/server/app-render/csrf-protection.ts]: Good catch! Implemented custom `asciiLowerCase()` that only modifies A-Z (65-90), leaving all other characters unchanged per RFC 1035.

Added tests for unicode/punycode handling. Also updated `baseline-browser-mapping` to 2.9.19 to fix the test failures.

File: packages/next/src/server/app-render/csrf-protection.test.ts [hunk 2]
```diff
@@ -77,4 +77,40 @@ describe('isCsrfOriginAllowed', () => {
...
+    
+    // Unicode characters should NOT be lowercased (security requirement)
+    // If someone passes raw unicode, it should be matched exactly
+    expect(isCsrfOriginAllowed('münchen.com', ['münchen.com'])).toBe(true)
+    expect(isCsrfOriginAllowed('MÜNCHEN.com', ['münchen.com'])).toBe(false) // Capital Ü stays capital
+    expect(isCsrfOriginAllowed('münchen.com', ['MÜNCHEN.com'])).toBe(false) // They must match exactly
+    
+    // ASCII portion should still be case-insensitive
+    expect(isCsrfOriginAllowed('münchen.COM', ['münchen.com'])).toBe(true)
+    expect(isCsrfOriginAllowed('MÜNCHEN.COM', ['MÜNCHEN.com'])).toBe(true)
```
Comment (bgw) [inline packages/next/src/server/app-render/csrf-protection.test.ts]: I think this does a better job of testing what the comment suggests:

```suggestion
    // ASCII portion should still be case-insensitive
    expect(isCsrfOriginAllowed('MüNCHEN.COM', ['münchen.com'])).toBe(true)
    expect(isCsrfOriginAllowed('mÜnchen.COM', ['MÜNCHEN.com'])).toBe(true)
```

File: packages/next/package.json [hunk 3]
```diff
@@ -99,7 +99,7 @@
   "dependencies": {
     "@next/env": "16.2.0-canary.13",
     "@swc/helpers": "0.5.15",
-    "baseline-browser-mapping": "^2.8.3",
+    "baseline-browser-mapping": "^2.9.19",
```
Comment (bgw) [inline packages/next/package.json:102]: Revert this. The CI failures should be fixed by https://github.com/vercel/next.js/pull/89175

**Review by bgw — CHANGES_REQUESTED** [review]
Regarding the trailing dot: After some further research, it does look like browsers do sometimes treat it as a separate, non-equivalent, hostname:

> The example.com and example.com. [domains](https://url.spec.whatwg.org/#concept-domain) are not equivalent and typically treated as distinct.

-- https://url.spec.whatwg.org/#concept-host

Though apparently the whole spec is a mess:
 
Image

Because this is security critical code where being overly restrictive isn't a security issue, but not being restrictive enough could result in a CVE, let's be conservative treat the trailing dot as a separate host/domain, like browsers do.

I don't see any practical way the trailing dot could be abused (and I had a chat with Claude to double-check), but I also don't think anyone was running into problems with that. I can believe people were running into case sensitivity issues, but that's more straightforward because all the specs are clear and agree about ASCII case-insensitive matching.


---

> Would you like me to update baseline-browser-mapping to the latest version as part of this PR, or should this be handled separately?

No thanks, just rebase on top of canary. https://github.com/vercel/next.js/pull/89175 should've fixed that issue.

**Comment by bgw** [comment]
Looks like you need to run `pnpm prettier-fix` or `pnpm prettier --write packages/next/src/server/app-render/csrf-protection.test.ts`

**Comment by jaffarkeikei** [comment]
Updated PR based on feedback:

✅ Rebased on latest canary
✅ Removed trailing dot handling (keeping conservative approach)
✅ Kept case-insensitive matching with ASCII-only character replacement
✅ Removed trailing dot test cases

The PR now only addresses case-insensitive DNS name matching per RFC 1035, using ASCII-only character replacement to avoid unicode issues.

**Review by bgw — APPROVED** [review]

**Comment by bgw** [comment]
Thanks for the contribution, @jaffarkeikei! Sorry about all the extra back-and-forth. Because this is code is security-related, I had to be careful that we're not screwing anything up.
